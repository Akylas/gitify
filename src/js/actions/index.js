import { apiRequest, apiRequestAuth } from '../utils/api-requests';

import Constants from '../utils/constants';

export function makeAsyncActionSet(actionName) {
  return {
    REQUEST: actionName + '_REQUEST',
    SUCCESS: actionName + '_SUCCESS',
    FAILURE: actionName + '_FAILURE'
  };
}

// Authentication

export const LOGIN = makeAsyncActionSet('LOGIN');
export function loginUser(code) {
  return (dispatch, getState) => {

    const url = 'https://github.com/login/oauth/access_token';
    const method = 'POST';
    const data = {
      'client_id': Constants.CLIENT_ID,
      'client_secret': Constants.CLIENT_SECRET,
      'code': code
    };

    dispatch({type: LOGIN.REQUEST});

    return apiRequest(url, method, data)
      .then(function (response) {
        dispatch({type: LOGIN.SUCCESS, payload: response.data});
      })
      .catch(function (error) {
        dispatch({type: LOGIN.FAILURE, payload: error.response.data});
      });

  };
};

export const LOGOUT = 'LOGOUT';
export function logout() {
  return { type: LOGOUT };
};


// Notifications

export const NOTIFICATIONS = makeAsyncActionSet('NOTIFICATIONS');
export function fetchNotifications() {
  return (dispatch, getState) => {

    const participating = getState().settings.get('participating');
    const url = `https://api.github.com/notifications?participating=${participating}`;
    const method = 'GET';
    const token = getState().auth.get('token');

    dispatch({type: NOTIFICATIONS.REQUEST});

    return apiRequestAuth(url, method, token)
      .then(function (response) {
        dispatch({type: NOTIFICATIONS.SUCCESS, payload: response.data});
      })
      .catch(function (error) {
        dispatch({type: NOTIFICATIONS.FAILURE, payload: error.response.data});
      });

  };
};


// Single Notification

export const MARK_NOTIFICATION = makeAsyncActionSet('MARK_NOTIFICATION');
export function markNotification(id) {
  return (dispatch, getState) => {

    const url = `https://api.github.com/notifications/threads/${id}`;
    const method = 'PATCH';
    const token = getState().auth.get('token');

    dispatch({type: MARK_NOTIFICATION.REQUEST});

    return apiRequestAuth(url, method, token, {})
      .then(function (response) {
        dispatch({type: MARK_NOTIFICATION.SUCCESS, payload: response.data, meta: { id }});
      })
      .catch(function (error) {
        dispatch({type: MARK_NOTIFICATION.FAILURE, payload: error.response.data});
      });
  };
};


// Repo's Notification

export const MARK_REPO_NOTIFICATION = makeAsyncActionSet('MARK_REPO_NOTIFICATION');
export function markRepoNotifications(loginId, repoId, repoFullName) {
  return (dispatch, getState) => {

    const url = `https://api.github.com/repos/${loginId}/${repoId}/notifications`;
    const method = 'PUT';
    const token = getState().auth.get('token');

    dispatch({type: MARK_REPO_NOTIFICATION.REQUEST});

    return apiRequestAuth(url, method, token, {})
      .then(function (response) {
        dispatch({type: MARK_REPO_NOTIFICATION.SUCCESS, payload: response.data, meta: { repoFullName, repoId }});
      })
      .catch(function (error) {
        dispatch({type: MARK_REPO_NOTIFICATION.FAILURE, payload: error.response.data});
      });
  };
};


// Starred

export const HAS_STARRED = makeAsyncActionSet('HAS_STARRED');
export function checkHasStarred() {
  return (dispatch, getState) => {

    const url = `https://api.github.com/user/starred/${Constants.REPO_SLUG}`;
    const method = 'GET';
    const token = getState().auth.get('token');

    dispatch({type: HAS_STARRED.REQUEST});

    return apiRequestAuth(url, method, token)
      .then(function (response) {
        dispatch({type: HAS_STARRED.SUCCESS, payload: response.data});
      })
      .catch(function (error) {
        dispatch({type: HAS_STARRED.FAILURE, payload: error.response.data});
      });

  };
};


// Search

export const SEARCH_NOTIFICATIONS = 'SEARCH_NOTIFICATIONS';
export const CLEAR_SEARCH = 'CLEAR_SEARCH';
export function searchNotifications(query) {
  return {
    type: SEARCH_NOTIFICATIONS,
    query: query
  };
};

export function clearSearch() {
  return {
    type: CLEAR_SEARCH
  };
};


// Settings

export const UPDATE_SETTING = 'UPDATE_SETTING';
export function updateSetting(setting, value) {
  return {
    type: UPDATE_SETTING,
    setting: setting,
    value: value
  };
};
