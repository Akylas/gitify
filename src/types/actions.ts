export const TOGGLE_SETTINGS_MODAL = 'TOGGLE_SETTINGS_MODAL';

export interface ToggleSettingsModalAction {
  type: typeof TOGGLE_SETTINGS_MODAL;
}
