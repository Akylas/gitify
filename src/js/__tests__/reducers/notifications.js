import { Map, List, fromJS } from 'immutable';

import reducer from '../../reducers/notifications';
import { NOTIFICATIONS, MARK_NOTIFICATION, MARK_REPO_NOTIFICATION } from '../../actions';

describe('reducers/notifications.js', () => {
  const initialState = Map({
    response: List(),
    isFetching: false,
    failed: false
  });

  const notifications = fromJS([
    {
      id: 1,
      repository: {
        full_name: 'manosim/gitify'
      },
      text: 'New Release'
    },
    {
      id: 2,
      repository: {
        full_name: 'manosim/gitify'
      },
      text: 'It\'s Great'
    }
  ]);

  it('should return the initial state', () => {

    expect(reducer(undefined, {})).toEqual(initialState);

  });

  it('should handle NOTIFICATIONS.REQUEST', () => {

    const action = {
      type: NOTIFICATIONS.REQUEST
    };

    expect(reducer(undefined, action)).toEqual(
      initialState
        .set('isFetching', true)
        .set('failed', false)
    );

  });

  it('should handle NOTIFICATIONS.SUCCESS', () => {

    expect(reducer(undefined, {})).toEqual(initialState);

    const action = {
      type: NOTIFICATIONS.SUCCESS,
      payload: notifications
    };

    const currentState =
      initialState
        .set('isFetching', true)
        .set('failed', false);

    expect(reducer(currentState, action)).toEqual(
      initialState
        .set('isFetching', false)
        .set('response', notifications)
    );

  });

  it('should handle NOTIFICATIONS.FAILURE', () => {

    const response = {
      error: 404,
      message: 'Oops! Something went wrong.'
    };

    const currentState =
      initialState
        .set('isFetching', true)
        .set('failed', false);

    expect(reducer(currentState, {})).toEqual(currentState);

    const action = {
      type: NOTIFICATIONS.FAILURE,
      payload: response
    };

    expect(reducer(currentState, action)).toEqual(
      initialState
        .set('isFetching', false)
        .set('failed', true)
        .set('response', List())
    );

  });

  it('should handle MARK_NOTIFICATION.SUCCESS', () => {

    const currentState = initialState.set('response', notifications);

    expect(reducer(currentState, {}).get('response').size).toBe(2);

    const action = {
      type: MARK_NOTIFICATION.SUCCESS,
      meta: {
        id: 1
      }
    };

    expect(reducer(currentState, action).get('response').size).toBe(1);

  });

  it('should handle MARK_REPO_NOTIFICATION.SUCCESS', () => {

    const currentState = initialState.set('response', notifications);

    expect(reducer(currentState, {}).get('response').size).toBe(2);

    const action = {
      type: MARK_REPO_NOTIFICATION.SUCCESS,
      meta: {
        repoSlug: 'manosim/gitify'
      }
    };

    expect(reducer(currentState, action).get('response').size).toBe(0);

  });
});
