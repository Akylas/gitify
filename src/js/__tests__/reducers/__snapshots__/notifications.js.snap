// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/notifications.js should handle NOTIFICATIONS.SUCCESS (github.com - empty list) 1`] = `
Object {
  "failed": false,
  "isFetching": false,
  "response": Array [
    Object {
      "hostname": "github.com",
      "notifications": Array [
        Object {
          "id": 1,
          "repository": Object {
            "full_name": "manosim/gitify",
          },
          "text": "New Release",
        },
        Object {
          "id": 2,
          "repository": Object {
            "full_name": "manosim/gitify",
          },
          "text": "It's Great",
        },
      ],
    },
  ],
}
`;

exports[`reducers/notifications.js should handle NOTIFICATIONS.SUCCESS (github.com - not empty list) 1`] = `
Object {
  "failed": false,
  "isFetching": false,
  "response": Array [
    Object {
      "hostname": "github.com",
      "notifications": Array [
        Object {
          "id": 1,
          "repository": Object {
            "full_name": "manosim/gitify",
          },
          "text": "New Release",
        },
        Object {
          "id": 2,
          "repository": Object {
            "full_name": "manosim/gitify",
          },
          "text": "It's Great",
        },
      ],
    },
  ],
}
`;
